moveit_ros/moveit_servo/tests/servo_cpp_fixture.hpp

```diff
@@ -61,23 +61,10 @@ class ServoCppFixture : public testing::Test
     servo_params_ = servo_param_listener_->get_params();
 
     planning_scene_monitor_ = moveit_servo::createPlanningSceneMonitor(servo_test_node_, servo_params_);
-
     // Wait until the joint configuration is nonzero before starting MoveIt Servo.
-    int num_tries = 0;
-    const int max_tries = 20;
-    while (true)
+    if (!planning_scene_monitor_->getStateMonitor()->waitForCompleteState("panda_arm", 2.0))
```

Comment:
As you can see, using the state monitor here failed the test

Reply:
I don't think the `StateMonitor` is the culprit here. Eventually, the `StateMonitor` receives joint state updates and _forwards_ those to the `PlanningScene`:
https://github.com/moveit/moveit2/blob/b3b1f73b3295abc14f1cd7dbbe851fa594b3b58b/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp#L1408-L1409
That is, if the PlanningScene reports the correct state, the StateMonitor should do so as well.

However, there is a key difference between the proposed code and yours: You are explicitly waiting for a _non-zero_ state, while `waitForCompleteState` should just wait for all group joints to be received once. Either the logics of `haveCompleteState()` is broken or some publisher provides zero joint states initially. Should be easy to figure out with some debug prints in `PlanningSceneMonitor::onStateUpdate()`.

Reply:
I printed the robot configuration after the call to `waitForCompleteState()` and always got nonzero (correct) robot states after when testing locally. 

Maybe then the state monitor gets a good robot state, but because the collision monitor is getting it from the planning scene, *that one* is not yet up to date? And we should actually be checking for a current state from the planning scene itself?

https://github.com/moveit/moveit2/blob/main/moveit_ros/moveit_servo/src/collision_monitor.cpp#L111-L115

Reply:
Confirming this is the case by adding this code after your `waitForCompleteState()`.

```cpp
    auto after = planning_scene_monitor_->getStateMonitor()->getCurrentState()->getVariablePositions();
    
    planning_scene_monitor::LockedPlanningSceneRO locked_scene(planning_scene_monitor_);
    auto after2 = locked_scene->getCurrentState().getVariablePositions();

    RCLCPP_ERROR_STREAM(servo_test_node_->get_logger(),
                        "[SCASTRO] STATE MONITOR: " << after[7] << " " << after[8] << " " << after[9] << " "
                                                    << after[10] << " " << after[11] << " " << after[12] << " "
                                                    << after[13]);

    RCLCPP_ERROR_STREAM(servo_test_node_->get_logger(),
                        "[SCASTRO] LOCKED PLANNING SCENE: " << after2[7] << " " << after2[8] << " " << after2[9] << " "
                                                            << after2[10] << " " << after2[11] << " " << after2[12]
                                                            << " " << after2[13]);
```

Result:
```
[moveit_servo_test]: [SCASTRO] STATE MONITOR: 0 -0.785 0 -2.356 0 1.571 0.785
[moveit_servo_test]: [SCASTRO] LOCKED PLANNING SCENE: 0 0 0 0 0 0 0
```

Reply:
I suggest going back to my implementation in my previous PR, but instead of returning the `Eigen::Map` we return the `std::vector` and do the Eigen map in the fixture setup so it does not go out of scope?

Reply:
> Is there an equivalent call that does not use the state monitor

Yes, there is [`PlanningSceneMonitor::waitForCurrentRobotState()`](https://github.com/moveit/moveit2/blob/b3b1f73b3295abc14f1cd7dbbe851fa594b3b58b/moveit_ros/planning/planning_scene_monitor/src/planning_scene_monitor.cpp#L1085-L1086). This method ensures a PS update after calling StateMonitor's `waitForCurrentState()`.
There is one downside: The method doesn't allow to specify a planning group. But for the tests that should be fine as there is only a single JSP.

Reply:
> There is one downside: The method doesn't allow to specify a planning group.

Returning to StateMonitor and explicitly updating the PlanningScene fixes that issue too.